src/client/dfuse/ops/read.c

```diff
@@ -102,6 +102,334 @@ dfuse_readahead_reply(fuse_req_t req, size_t len, off_t position, struct dfuse_o
 	return true;
 }
 
+static struct dfuse_eq *
+pick_eqt(struct dfuse_info *dfuse_info)
+{
+	uint64_t eqt_idx;
+
+	eqt_idx = atomic_fetch_add_relaxed(&dfuse_info->di_eqt_idx, 1);
+	return &dfuse_info->di_eqt[eqt_idx % dfuse_info->di_eq_count];
+}
+
+/* Chunk read and coalescing
+ *
+ * This code attempts to predict application and kernel I/O patterns and preemptively read file
+ * data ahead of when it's requested.
+ *
+ * For some kernels read I/O size is limited to 128k when using the page cache or 1Mb when using
+ * direct I/O.  To get around the performance impact of them detect when well aligned 128k reads
+ * are received and read an entire buffers worth, then for future requests the data should already
+ * be in cache.
+ *
+ * This code is entered when caching is enabled and reads are correctly size/aligned and not in the
```

Comment:
which test is using this new feature?

Reply:
Here's the result from a test on this PR:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14212/23/artifact/Functional%20Hardware%20Medium/dfuse/caching_check.py/
And the same result against master:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/380/artifact/Functional%20Hardware%20Medium/dfuse/caching_check.py/

You can see the read performance in the forth iteration goes from `1317.47` to `3087.97`MiB/s

dfuse can deliver wire-speed on reads but this change will improve the single-client performance and the in-node scaling so higher bandwidth figures will be available with smaller client counts.

Reply:
im not questioning your conclusion that this does provide a better result, but I think you know and you are the one who always says that CI is not valid for looking at performance numbers ;-)

Reply:
You're right that this is a major frustration/limitation with CI.  Once you understand the kernel behaviour though and look at what I/O is being requested by dfuse then the potential should be obvious.  I also see performance improvements locally but that's harder to link to here and whilst we can't trust CI numbers entirely a > 100% improvement in bandwidth is not to be disregarded, even if there are large error bars on that.

Reply:
again, im not disputing your assessment. 
But i do want to point something obvious that you are missing here. In many applications and middleware on top of POSIX, usually there is always an offset or metadata in the file that breaks that 128k alignment. so "well-behaved" IO can be large and contiguous, but in most cases not 128k aligned.
I did not block your PR here because this is not hurting anything really, but i doubt this provides much benefit beyond some ior runs that are 128k aligned.

Reply:
This absolutely won't catch everyone, but it should help for all applications that have put the effort into optimising their I/O patterns.  Lustre for example has been recommending MB size/ MB aligned I/Os for twenty years now.

The more flexible we make this the more applications will potentially benefit but also the greater the cost of incorrectly triggering which will lead to wasted I/O which would be more cost for less benefit.  I view this as a conservative but safe set of requirements today that we can extend, once we have mechanisms in place in order to allow tuning to avoid overburden servers.

Reply:
lustre can recommend anything, that does not mean in reality this is what is happening.

Reply:
and im definitely disputing your claim that optimizing IO pattern == making it 128k aligned.. this is definitely false IMO